### PR TITLE
Add OAuth2 authentication server

### PR DESCRIPTION
**Pull request by sbrunner**
GEO-4082

See reference image https://oauthlib.readthedocs.io/en/latest/_images/graphviz-70406cc7e2f849ca07fa1b49bd62b208d0ff3278.png :-)

### CODE REVIEW

File: geoportal/c2cgeoportal_geoportal/__init__.py [hunk 1]
```diff
@@ -207,6 +207,23 @@ def get_user_from_request(request, username=None):
...
+                valid, oauth2_request = oauth2.get_oauth_client().verify_request(
+                    request.current_route_url(_query=request.GET),
+                    request.method,
+                    request.body,
+                    request.headers,
+                    [],
+                )
+                LOG.info("OAuth verify_request: %s", valid)
+                if valid:
+                    request.user_ = oauth2_request.user
```
Comment (arnaud-morvan) [inline geoportal/c2cgeoportal_geoportal/__init__.py:226]: It might be more flexible to create an authentication policy.

Comment (sbrunner) [reply geoportal/c2cgeoportal_geoportal/__init__.py:226]: +1 Thanks :-)

File: geoportal/c2cgeoportal_geoportal/lib/oauth2.py [hunk 2]
```diff
@@ -0,0 +1,965 @@
...
+        """
+        del args, kwargs
+
+        LOG.debug("save_authorization_code %s", client_id)
+
+        from c2cgeoportal_commons.models import DBSession, static  # pylint: disable=import-outside-toplevel
+
+        user = pyramid.threadlocal.get_current_request().user_
+
+        # Don't allows to have tow authentication for one user end one client
```
Comment (arnaud-morvan) [inline geoportal/c2cgeoportal_geoportal/lib/oauth2.py:507]:
```suggestion
        # Don't allows to have two authentications for the same user and the same client
```

File: geoportal/c2cgeoportal_geoportal/lib/oauth2.py [hunk 3]
```diff
@@ -0,0 +1,965 @@
...
+            - Resource Owner Password Credentials Grant (might not associate a client)
+            - Client Credentials grant
+        """
+        del args, kwargs
+
+        LOG.debug("save_bearer_token")
+
+        from c2cgeoportal_commons.models import DBSession, static  # pylint: disable=import-outside-toplevel
+
+        # Don't allows to have tow token for one user end one client
```
Comment (arnaud-morvan) [inline geoportal/c2cgeoportal_geoportal/lib/oauth2.py:577]:
```suggestion
        # Don't allows to have two tokens for the same user and the same client
```

File: geoportal/c2cgeoportal_geoportal/lib/oauth2.py [hunk 4]
```diff
@@ -0,0 +1,965 @@
...
+            - Client Credentials Grant
+        """
+
+        LOG.debug("validate_bearer_token %s", scopes)
+
+        from c2cgeoportal_commons.models import DBSession, static  # pylint: disable=import-outside-toplevel
+
+        bearer_token = (
+            DBSession.query(static.OAuth2BearerToken).filter(static.OAuth2BearerToken.access_token == token)
+            # .filter(static.OAuth2BearerToken.client_id == request.client.client_id)
```
Comment (arnaud-morvan) [inline geoportal/c2cgeoportal_geoportal/lib/oauth2.py:660]: Forgotten code comment ?

File: geoportal/c2cgeoportal_geoportal/scaffolds/update/geoportal/CONST_vars.yaml_tmpl [hunk 5]
```diff
@@ -40,6 +40,8 @@ vars:
   authentication:
     two_factor: false
     two_factor_issuer_name: {{project}}
+    oauth2_token_expire_minutes: 60
+    oauth2_authorization_expire_minutes: 10
```
Comment (arnaud-morvan) [inline geoportal/c2cgeoportal_geoportal/scaffolds/update/geoportal/CONST_vars.yaml_tmpl:44]: Are you sure about these values, I would think the token expire after 10 minutes but the authorization remains for 60 min.

File: geoportal/c2cgeoportal_geoportal/views/login.py [hunk 6]
```diff
@@ -155,15 +166,47 @@ def login(self):
...
+                    if body:
+                        raise exception_response(status, details=body)
+                    raise exception_response(status)
+                return set_common_headers(
+                    self.request,
+                    "login",
+                    NO_CACHE,
+                    response=Response(body, headers=headers.items()),
+                )
+            headers = remember(self.request, username)
```
Comment (arnaud-morvan) [inline geoportal/c2cgeoportal_geoportal/views/login.py:209]: This is a huge function now. Hope we will be able to cut it at some time.
Won't it make sense to create a dedicated enpoint for OAuth ?

File: geoportal/tests/functional/test_oauth2.py [hunk 7]
```diff
@@ -0,0 +1,253 @@
...
+        from c2cgeoportal_commons.models import DBSession
+        from c2cgeoportal_commons.models.static import User
+        from c2cgeoportal_geoportal.views.login import Login
+
+        # Test login form
+        request = create_dummy_request()
+        request.params = {
+            "client_id": "qgis",
+            "redirect_uri": "http://127.0.0.1:7070/",
+            "responce_type": "token",
```
Comment (arnaud-morvan) [inline geoportal/tests/functional/test_oauth2.py:90]:
```suggestion
            "response_type": "token",
```

File: geoportal/tests/functional/test_oauth2.py [hunk 8]
```diff
@@ -0,0 +1,253 @@
...
+        }
+        response = Login(request).oauth2loginform()
+        assert response["login_params"] == {
+            "client_id": "qgis",
+            "redirect_uri": "http://127.0.0.1:7070/",
+            "responce_type": "token",
```
Comment (arnaud-morvan) [inline geoportal/tests/functional/test_oauth2.py:96]:
```suggestion
            "response_type": "token",
```

**Review by arnaud-morvan — APPROVED** [review]
Minor comments, trust you on the OAuth2 functional part.

File: geoportal/c2cgeoportal_geoportal/lib/oauth2.py (hunk 4 above)
Comment (sbrunner) [reply geoportal/c2cgeoportal_geoportal/lib/oauth2.py:660]: Thanks

File: geoportal/c2cgeoportal_geoportal/scaffolds/update/geoportal/CONST_vars.yaml_tmpl (hunk 5 above)
Comment (sbrunner) [reply geoportal/c2cgeoportal_geoportal/scaffolds/update/geoportal/CONST_vars.yaml_tmpl:44]: It's correct, I use the proposal values from here: https://oauthlib.readthedocs.io/en/latest/oauth2/server.html

File: admin/c2cgeoportal_admin/schemas/oauth2_clients.py [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+
+
+import colander
+from c2cgeoform.ext.deform_ext import RelationCheckBoxListWidget
+from c2cgeoform.schema import GeoFormManyToManySchemaNode, manytomany_validator
+
+from c2cgeoportal_admin import _
+from c2cgeoportal_commons.models.static import OAuth2Client
+
+oauth2clients_schema_node = colander.SequenceSchema(
```
[the remark this answers is not in the document]
Comment (sbrunner) [reply admin/c2cgeoportal_admin/schemas/oauth2_clients.py:38]: I don't know c2cgeoform well, but if you say that I can remove it I will do it...

File: geoportal/c2cgeoportal_geoportal/lib/authentication.py [hunk 10]
```diff
@@ -95,6 +96,34 @@ def forget(self, request):  # pylint: disable=unused-argument, no-self-use
...
+        valid, oauth2_request = oauth2.get_oauth_client(request.registry.settings).verify_request(
+            route_url,
+            request.method,
+            request.body,
+            request.headers,
+            [],
+        )
+        LOG.info("OAuth verify_request: %s", valid)
+        if valid:
+            request.user_ = oauth2_request.user
```
[the remark this answers is not in the document]
Comment (sbrunner) [reply geoportal/c2cgeoportal_geoportal/lib/authentication.py:124]: +1 effectively

File: geoportal/c2cgeoportal_geoportal/views/login.py (hunk 6 above)
Comment (sbrunner) [reply geoportal/c2cgeoportal_geoportal/views/login.py:209]: I extract the function, I'm not certain if that's a good way to create a dedicated enpoint, I don't want to duplicate code for that...

**Comment by arnaud-morvan** [comment]
@sbrunner : Note that this PR does not add any tests on admin part.

**Pull request merged**